pkg/performance/collectors/kernel.go

```diff
@@ -0,0 +1,448 @@
+// Reference: https://www.kernel.org/doc/Documentation/ABI/testing/dev-kmsg
+type KernelCollector struct {
+	performance.BaseCollector
+	logger       logr.Logger
+	kmsgPath     string
+	messageLimit int
+	procUtils    *procutils.ProcUtils
+
+	// Continuous collection state
+	continuousMu     sync.Mutex
+	continuousCtx    context.Context
+	continuousCancel context.CancelFunc
+	continuousChan   chan any
+	isRunning        bool
+	lastError        error
+}
+
+type KernelCollectorOption func(*KernelCollector)
+
+func WithMessageLimit(limit int) KernelCollectorOption {
+	return func(c *KernelCollector) {
+		if limit > 0 {
+			c.messageLimit = limit
+		}
+	}
+}
+
+func NewKernelCollector(logger logr.Logger, config performance.CollectionConfig, opts ...KernelCollectorOption) *KernelCollector {
+	capabilities := performance.CollectorCapabilities{
+		SupportsOneShot:    true,
+		SupportsContinuous: true,
+		RequiresRoot:       true, // /dev/kmsg typically requires CAP_SYSLOG or root
+		RequiresEBPF:       false,
+		MinKernelVersion:   "3.5.0", // /dev/kmsg was introduced in 3.5
+	}
+
+	collector := &KernelCollector{
+		BaseCollector: performance.NewBaseCollector(
+			performance.MetricTypeKernel,
+			"Kernel Message Collector",
+			logger,
+			config,
+			capabilities,
+		),
+		logger:       logger.WithName("kernel"),
+		kmsgPath:     filepath.Join(config.HostDevPath, "kmsg"),
+		messageLimit: defaultMessageLimit,
+		procUtils:    procutils.New(config.HostProcPath),
+	}
+
+	for _, opt := range opts {
+		opt(collector)
+	}
+
+	return collector
+}
+
+func (c *KernelCollector) Collect(ctx context.Context) (any, error) {
+	messages, err := c.collectKernelMessages(ctx)
+	if err != nil {
+		if os.IsPermission(err) {
+			c.logger.V(1).Info("Permission denied reading kernel messages", "path", c.kmsgPath)
+			return []*performance.KernelMessage{}, nil
+		}
+		return nil, err
+	}
+	return messages, nil
+}
+
+func (c *KernelCollector) collectKernelMessages(ctx context.Context) ([]*performance.KernelMessage, error) {
+	bootTime, err := c.procUtils.GetBootTime()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get boot time: %w", err)
+	}
+
+	// Open /dev/kmsg with O_RDONLY | O_NONBLOCK
+	// O_NONBLOCK is critical to avoid blocking when no messages are available
+	fd, err := syscall.Open(c.kmsgPath, syscall.O_RDONLY|syscall.O_NONBLOCK, 0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open %s: %w", c.kmsgPath, err)
+	}
+	defer syscall.Close(fd)
+
+	// Note: We don't seek to the end because:
+	// 1. Each reader of /dev/kmsg gets their own view of the ring buffer
+	// 2. The first read will give us the oldest available message
+	// 3. We limit the number of messages returned via messageLimit
+
+	// Ring buffer to store raw message strings
+	ringBuf, err := ringbuffer.New[string](c.messageLimit)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ring buffer: %w", err)
+	}
+	buf := make([]byte, kmsgBufferSize)
```

Comment:
I'd imagine a future optimization would be to make multiple collectors share the same ring buffer?

Reply:
I'm not sure that would be really necessary. The only reason we are using a ring buffer in userspace here is to efficiently calculate the last-N messages as we read from /dev/kmsg (which is a kernel-space ring buffer itself... funily enough).